### PR TITLE
feat: support for opening/closing private broker channels

### PR DESCRIPTION
**Pull request by msgmaxim**
# Pull Request

Closes: PRO-1748

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have written sufficient tests.

## Summary

- Added a storage map `AccountId` -> `ChannelId`
- Added Extrinsics to open/close private channels + tests + benchmarks

Out of scope (these seem like may need some further discussion):
- Charging fee/rent for opening a private channel/keeping it open
- Closing private channels upon account deletion (maybe this isn't necessary for an MVP?)
(will create separate linear tickets for these if we agree to leave them out of this PR)

May need to rebase this after #5357 is merged.

### CODE REVIEW

File: state-chain/pallets/cf-ingress-egress/src/lib.rs
Comment (dandanlen) [inline state-chain/pallets/cf-ingress-egress/src/lib.rs:587]: This can use the Identity hasher.

```suggestion
		StorageMap<_, Identity, T::AccountId, ChannelId, OptionQuery>;
```

File: state-chain/pallets/cf-ingress-egress/src/lib.rs
Comment (kylezs) [inline state-chain/pallets/cf-ingress-egress/src/lib.rs:1345]: we should make this a function, is duplicate

File: state-chain/pallets/cf-ingress-egress/src/lib.rs
Comment (kylezs) [inline state-chain/pallets/cf-ingress-egress/src/lib.rs:1336]: Should we also restrict this to be just for Bitcoin - it's not useful for any other chains right? probably more importantly is that not restricting it to Bitcoin would be confusing when reading the code, since it's an instantiable pallet, we'd assume it's for all instances.

File: state-chain/pallets/cf-ingress-egress/src/lib.rs
Comment (kylezs) [inline state-chain/pallets/cf-ingress-egress/src/lib.rs:1340]: I see this burning/rent is mentioned in the ticket, but also that this PR closes that ticket? so maybe we create another to address this separately.

**Comment by dandanlen** [comment]
> Closing private channels upon account deletion (maybe this isn't necessary for an MVP?)
(will create separate linear tickets for these if we agree to leave them out of this PR)

I feel like we can just add this here. We need it and it's not a lot of work to add. Not sure what we gain by waiting. 

(Sorry was sure I left this comment yesterday...)

File: state-chain/pallets/cf-ingress-egress/src/lib.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-ingress-egress/src/lib.rs:1340]: I will create tickets once we decide what should be in the scope of this PR and what isn't (I mentioned this in the description).

Comment (kylezs) [reply state-chain/pallets/cf-ingress-egress/src/lib.rs:1340]: > Closing private channels upon account deletion (maybe this isn't necessary for an MVP?)
(will create separate linear tickets for these if we agree to leave them out of this PR)

Ah ok, that seemed to suggest was just about the account deletion

**Comment by dandanlen** [comment]
Forgot to mention - I think charging/not charging a fee needs a little more discussion, happy to leave this out for now.

File: state-chain/pallets/cf-ingress-egress/src/lib.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-ingress-egress/src/lib.rs]: Deduplicated

File: state-chain/pallets/cf-ingress-egress/src/lib.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-ingress-egress/src/lib.rs]: Added the check.

**Comment by msgmaxim** [comment]
Addressed all comments now. Ready for re-review @dandanlen @kylezs

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (kylezs) [inline state-chain/pallets/cf-swapping/src/lib.rs:1084]: Now we only inject the BitcoinIngressEgress as the PrivateChannelManager, so it's only possible to create bitcoin channel, or inversely, if we injected the wrong instance of the ingress egress then we would put bitcoin in here, and then open an Ethereum channel. 
 So perhaps would be better to have this be:
```fn open_private_btc_channel(origin)``` and we can remove this ensure, same below.

File: state-chain/pallets/cf-swapping/src/tests.rs
Comment (kylezs) [inline state-chain/pallets/cf-swapping/src/tests.rs:1690]:
```suggestion
			// Only brokers can open channels
```

**Review by kylezs — APPROVED** [review]
Nice tests 👌  - would let Dan approve before merge

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [reply state-chain/pallets/cf-swapping/src/lib.rs:1084]: Was going to suggest the same thing - the only benefit of the `chain` parameter is in case we want to add another utxo chain... but that seems unlikely right now and in the meantime it's not used for anything.

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [inline state-chain/pallets/cf-swapping/src/lib.rs:1098]: Same here, including the `chain` param seems unnecessary.

**Review by dandanlen — COMMENTED** [review]
I think we might as well make it btc-specific and remove the chain params to make it clearer.

**Comment by msgmaxim** [comment]
Renamed the extrinsics in cf-swapping and removed the chain parameter. Note that I chose not to rename the methods inside cf-ingress-egress as I thought it would look odd (it would be misleading for non-btc instances to have a btc-specific method), but I added a comment to say that this functionality is expected to be used by bitcoin only. Importantly, we don't have a chain parameter and we made it more explicit to external users that this is bitcoin only feature. @dandanlen

**Comment by kylezs** [comment]
> Note that I chose not to rename the methods inside cf-ingress-egress

I think that makes sense 👍

**Review by kylezs — APPROVED** [review]

**Review by dandanlen — COMMENTED** [review]
LGTM - only thing is that looking at the code now, i'm not sure why we put the storage in the ingess-egress pallet. All it needs to do is return an incremented channel id. We could store it in the swapping pallet and then we have a smaller internal API surface (no need for close/lookup methods in the trait). 

Will leave it up to you @msgmaxim whether to make this change or not. I think it would reduce complexity but having the 'channel' (even though it's just the id) in the same place as other channels also kind of makes sense.

**Comment by msgmaxim** [comment]
> LGTM - only thing is that looking at the code now, i'm not sure why we put the storage in the ingess-egress pallet. All it needs to do is return an incremented channel id. We could store it in the swapping pallet and then we have a smaller internal API surface (no need for close/lookup methods in the trait).
> 
> Will leave it up to you @msgmaxim whether to make this change or not. I think it would reduce complexity but having the 'channel' (even though it's just the id) in the same place as other channels also kind of makes sense.

Made these changes now @dandanlen

**Review by dandanlen — APPROVED** [review]

**Pull request merged**
